### PR TITLE
aws - add generic filter for security-hub findings

### PR DESCRIPTION
**Pull request by FireballDWF**
This PR is ready for review.

This addresses the finding filter part of https://github.com/cloud-custodian/cloud-custodian/issues/3191

### CODE REVIEW

File: c7n/filters/securityhub.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Copyright 2017 Capital One Services, LLC
```
Comment (kapilt) [inline c7n/filters/securityhub.py]: wrong year

File: c7n/filters/securityhub.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+from c7n.manager import resources
+
+
+class SecurityHubFindingFilter(Filter):
+    """Check if there are Security Hub Findings related to the resources
+    """
+    schema = type_schema('findings')
+    permissions = ('securityhub:GetFindings',)
+
+    def get_arn(self, r):
```
Comment (kapilt) [inline c7n/filters/securityhub.py]: use `self.manager.get_arns(resources)`

File: c7n/filters/securityhub.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+from c7n.utils import local_session, chunks, type_schema
+from .core import Filter
+from c7n.manager import resources
+
+
+class SecurityHubFindingFilter(Filter):
+    """Check if there are Security Hub Findings related to the resources
+    """
+    schema = type_schema('findings')
```
Comment (kapilt) [inline c7n/filters/securityhub.py]: while not universal we typically use singular form, ie `finding`

File: c7n/filters/securityhub.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    def get_arn(self, r):
+        # need to make generic to support other resource types
+        return "arn:aws:{}:{}:instance/{}".format(
+            self.manager.config.region,
+            self.manager.config.account_id,
+            r["InstanceId"])
+
+    def process(self, resources, event=None):
+        if not resources:
```
Comment (kapilt) [inline c7n/filters/securityhub.py]: this shouldn't be nesc, actions aren't invoked with empty resource sets.

File: tests/test_securityhub.py [hunk 5]
```diff
@@ -214,6 +214,34 @@ def test_instance(self):
...
+            },
+            config={"account_id": "101010101111"},
+            session_factory=factory,
+        )
+
+        resources = policy.run()
+        self.assertEqual(len(resources), 1)
+
+        client = factory().client("securityhub")
+        findings = client.get_findings(
```
Comment (kapilt) [inline tests/test_securityhub.py]: this api call isn't nesc for the test, there isn't any mutation side effect from a filter, the result of finding the resource, and asserting a filter added annotation is enough.

File: c7n/filters/securityhub.py (hunk 3 above)
Comment (kapilt) [inline c7n/filters/securityhub.py]: ideally, there should be a lot of filter options available for a user here that can pass through.

File: tests/test_securityhub.py (hunk 5 above)
Comment (FireballDWF) [reply tests/test_securityhub.py]: deleted

File: c7n/filters/securityhub.py (hunk 3 above)
Comment (FireballDWF) [reply c7n/filters/securityhub.py]: Agreed, captured in line:                 # TODO: Support parameterized filters rather not just finding exists

File: c7n/filters/securityhub.py (hunk 1 above)
Comment (FireballDWF) [reply c7n/filters/securityhub.py]: fixed

File: c7n/filters/securityhub.py (hunk 4 above)
Comment (kapilt) [reply c7n/filters/securityhub.py]: nope, we short-circuit execution once we get to an empty set on both filters and actions as there isn't anything for them to process.

File: c7n/filters/securityhub.py (hunk 3 above)
Comment (kapilt) [reply c7n/filters/securityhub.py]: yeah preference would be singular, just checked all the ec2 filters, and their all singular form.

```
custodian schema ec2.filters                                                                                                                                                                                                                                             
aws.ec2:
  filters: [and, config-compliance, default-vpc, ebs, ephemeral, event, health-event,
    image, image-age, instance-age, instance-attribute, instance-uptime, marked-for-op,
    metrics, network-location, not, offhour, onhour, or, security-group, singleton,
    state-age, subnet, tag-count, termination-protected, user-data, value, vpc]
```

Comment (kapilt) [reply c7n/filters/securityhub.py]: I think we'll need this out of the box within this pr, simply filtering to resources with any finding, doesn't seem per se useful without also being to narrow on specific types of findings.

File: c7n/filters/securityhub.py (hunk 4 above)
Comment (FireballDWF) [reply c7n/filters/securityhub.py]: ok, I've removed from securityhub code and will later submit to remove from line 41-42 of health.py which is the example I used as a starting point.

File: c7n/filters/securityhub.py (hunk 3 above)
Comment (FireballDWF) [reply c7n/filters/securityhub.py]: Agreed, I was trying to indicate that in my PR description.

File: c7n/filters/securityhub.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        'finding',
+        types={'type': 'array', 'items': {'type': 'string'}}
+    )
+    permissions = ('securityhub:GetFindings',)
+
+    def process(self, resources, event=None):
+        client = local_session(self.manager.session_factory).client(
+            'securityhub', region_name='us-east-1')
+
+        if self.manager.data['resource'] in {'app-elb'}:
```
Comment (kapilt) [inline c7n/filters/securityhub.py]: this filter feels like it started with copy paste from the health filter, which had a specific reasoning on the special treatment of alb, which isn't appropriate here.

Comment (FireballDWF) [reply c7n/filters/securityhub.py]: guilty as charged.  I refactor that out.

**Comment by kapilt** [comment]
The utility func for validation in #3358 can be used to validate the get findings filter as a passthrough

**Pull request merged**